src/Std/Sat/AIG/Basic.lean

```diff
@@ -0,0 +1,506 @@
+  | push_id wf ih =>
+    rw [Array.get_push]
+    split
+    . apply ih
+      simp [hfound]
+    . next hbounds =>
+      exfalso
+      apply hbounds
+      specialize ih _ hfound
+      apply Array.lt_of_getElem
+      assumption
+  | push_cache wf ih =>
+    rename_i decl'
+    rw [Array.get_push]
+    split
+    . simp only [HashMap.getElem?_insert] at hfound
+      match heq:decl == decl' with
+      | true =>
+        simp only [beq_iff_eq] at heq
+        simp [heq] at hfound
+        omega
+      | false =>
+        apply ih
+        simpa [BEq.symm_false heq] using hfound
+    . next hbounds =>
+      simp only [HashMap.getElem?_insert] at hfound
+      match heq:decl == decl' with
+      | true =>
+        apply Eq.symm
+        simpa using heq
+      | false =>
+        exfalso
+        apply hbounds
+        simp only [BEq.symm_false heq, cond_false] at hfound
+        specialize ih _ hfound
+        apply Array.lt_of_getElem
+        assumption
+
+/--
+Lookup a `Decl` in a `Cache`.
+-/
+opaque Cache.get? (cache : Cache α decls) (decl : Decl α) : Option (CacheHit decls decl) :=
+  /-
+  This function is marked as `opaque` to make sure it never, ever gets unfolded anywhere.
+  Unfolding it will often cause `HashMap.find?` to be symbolically evaluated by reducing
+  it either in `whnf` or in the kernel. This causes *huge* performance issues in practice.
+  The function can still be fully verified as all the proofs we need are in `CacheHit`.
+  -/
+  match hfound:cache.val[decl]? with
+  | some hit =>
+    some ⟨hit, Cache.get?_bounds _ _ hfound, Cache.get?_property _ _ hfound⟩
+  | none => none
+
+/--
+An `Array Decl` is a Direct Acyclic Graph (DAG) if a gate at index `i` only points to nodes with index lower than `i`.
+-/
+def IsDag (α : Type) (decls : Array (Decl α)) : Prop :=
+  ∀ {i lhs rhs linv rinv} (h : i < decls.size),
+      decls[i] = .gate lhs rhs linv rinv → lhs < i ∧ rhs < i
+
+/--
+The empty array is a DAG.
+-/
+theorem IsDag.empty {α : Type} : IsDag α #[] := by
+  intro i lhs rhs linv rinv h
+  simp only [Array.size_toArray, List.length_nil] at h
+  omega
+
+end AIG
+
+/--
+An And Inverter Graph (AIG α) together with a cache for subterm sharing.
+-/
+structure AIG (α : Type) [DecidableEq α] [Hashable α] where
+  /--
+  The circuit itself as an `Array Decl` whose members have indices into said array.
+  -/
+  decls : Array (AIG.Decl α)
+  /--
+  The `Decl` cache, valid with respect to `decls`.
+  -/
+  cache : AIG.Cache α decls
+  /--
+  In order to be a valid AIG, `decls` must form a DAG.
+  -/
+  inv : AIG.IsDag α decls
+
+namespace AIG
+
+/--
+An `AIG` with an empty AIG and cache.
+-/
+def empty : AIG α := { decls := #[], cache := Cache.empty #[], inv := IsDag.empty }
+
+/--
+The atom `a` occurs in `aig`.
+-/
+def Mem (a : α) (aig : AIG α) : Prop := (.atom a) ∈ aig.decls
+
+instance : Membership α (AIG α) where
+  mem := Mem
+
+/--
+A reference to a node within an AIG. This is the `AIG` analog of `Bool`.
```

Comment:
I don't understand the intent of "This is the `AIG` analog of `Bool`."

Reply:
`Bool` is a single bit and just like that an `Ref` is a single bit but encoded with the AIG. Similarly `BitVec` is a collection of bits and `RefVec` is a collection of bits that are encoded within the AIG.